notebooks/24-anomaly_detection.ipynb

```diff
@@ -0,0 +1,816 @@
+   "source": [
+    "fig, axes = plt.subplots(1, 5, figsize=(10, 4))\n",
+    "for ax, x in zip(axes, X_5[:5]):\n",
+    "    img = x.reshape(8, 8)\n",
+    "    ax.imshow(img, cmap=plt.cm.gray_r, interpolation='nearest')\n",
+    "    ax.axis('off')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "deletable": true,
+    "editable": true
+   },
+   "source": [
+    "1. Let's use IsolationForest to find the top 5% most abnormal images.\n",
+    "2. Let's plot them !"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "from sklearn.ensemble import IsolationForest\n",
+    "iforest = IsolationForest(contamination=0.05)\n",
+    "iforest = iforest.fit(X_5)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "deletable": true,
+    "editable": true
+   },
+   "source": [
+    "Compute the level of \"abnormality\" with `iforest.decision_function`. The lower, the more abnormal."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "iforest_X = iforest.decision_function(X_5)\n",
+    "plt.hist(iforest_X);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's plot the strongest inliers"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "X_strong_inliers = X_5[np.argsort(iforest_X)[-10:]]\n",
+    "\n",
+    "fig, axes = plt.subplots(2, 5, figsize=(10, 5))\n",
+    "\n",
+    "for i, ax in zip(range(len(X_outliers)), axes.ravel()):\n",
+    "    ax.imshow(X_strong_inliers[i].reshape((8, 8)),\n",
+    "               cmap=plt.cm.gray_r, interpolation='nearest')\n",
+    "    ax.axis('off')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's plot the strongest outliers"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "fig, axes = plt.subplots(2, 5, figsize=(10, 5))\n",
+    "\n",
+    "X_outliers = X_5[iforest.predict(X_5) == -1]\n",
+    "\n",
+    "for i, ax in zip(range(len(X_outliers)), axes.ravel()):\n",
+    "    ax.imshow(X_outliers[i].reshape((8, 8)),\n",
+    "               cmap=plt.cm.gray_r, interpolation='nearest')\n",
+    "    ax.axis('off')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-success\">\n",
+    "    <b>EXERCISE</b>:\n",
+    "     <ul>\n",
+    "      <li>\n",
+    "      Rerun the same analysis with all the other digits\n",
```

Comment:
Individually with a for-loop? Or all together? Or "for any given digit"?
Solution does "for any given digit", right?